### PR TITLE
bpo-23395: Raise an exception if the SIGINT signal is ignored or not handled

### PR DESCRIPTION
**Pull request by mcepl**
Attached interrupt_main.patch fixes for ``_thread.interrupt_main()``. Raise an exception if the ``SIGINT`` signal is ignored (``SIG_IGN``) or not handled by Python (``SIG_DFL``).

The patch updates the documentation and adds unit tests.

https://bugs.python.org/issue23395

### CODE REVIEW

**Comment by pablogsal** [comment]
Hi @mcepl and thank you for your contribution!

Notice that the CI is failing right now with this error:

```exception
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_threading.py", line 1194, in test_interrupt_main_error
    handler = signal.getsignal(signal.SIGINT)
NameError: name 'signal' is not defined
```

This is because you are using the `signal` module in the tests (`test_interrupt_main_error`) but you did not import it. Notice that to avoid waiting for the CI to fail in the PR you can run your tests locally. For example:

```
./configure --with-pydebug && make && make test
```

**Comment by pablogsal** [comment]
The tests keep failing:

```
======================================================================
ERROR: test_interrupt_main_error (test.test_threading.InterruptMainTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_threading.py", line 1199, in test_interrupt_main_error
    _thread.interrupt_main()
TypeError: 'int' object is not callable
----------------------------------------------------------------------
Ran 161 tests in 12.431s
```

Please, remember to run the tests locally to check that the patch works :)

**Comment by mcepl** [comment]
> Please, remember to run the tests locally to check that the patch works :)

This was marked as ``[WIP]`` as recording of the patch for [bug 23395](https://bugs.python.org/issue23395) by @vstinner . I thought it is better to have it here, so somebody who knows what's going on can take a look, than this patch be forever lost in bowls of the issue tracking system.

I personally don't understand why ``_thread.interrupt_main()`` which seems to me like a nice function, suddenly changes to ``int``, so I cannot help with fixing it.

**Comment by pablogsal** [comment]
The tests are not passing because you did not apply the patch completely. You missed

```
diff --git a/Modules/_threadmodule.c b/Modules/_threadmodule.c
index 69e27be9cc..d3ed1b9acf 100644
--- a/Modules/_threadmodule.c
+++ b/Modules/_threadmodule.c
@@ -1103,7 +1103,9 @@ thread to exit silently unless the exception is caught.");
 static PyObject *
 thread_PyThread_interrupt_main(PyObject * self, PyObject *Py_UNUSED(ignored))
 {
-    PyErr_SetInterrupt();
+    if (PyErr_SetInterruptWithErr() < 0) {
+    return NULL;
+    }
     Py_RETURN_NONE;
 }
```

you need to add the function to the public API:

```
diff --git a/Include/pyerrors.h b/Include/pyerrors.h
index 416d750d9b..5011f601d1 100644
--- a/Include/pyerrors.h
+++ b/Include/pyerrors.h
@@ -356,6 +356,7 @@ PyAPI_FUNC(PyObject *) _PyErr_TrySetFromCause(
 /* In signalmodule.c */
 PyAPI_FUNC(int) PyErr_CheckSignals(void);
 PyAPI_FUNC(void) PyErr_SetInterrupt(void);
+PyAPI_FUNC(int) PyErr_SetInterruptWithErr(void);

```

File: Modules/signalmodule.c
Comment (vstinner) [inline Modules/signalmodule.c]: You must hold the GIL to call PyErr_SetString().

Comment (mcepl) [reply Modules/signalmodule.c]: @vstinner That function seems to be called only in [thread_PyThread_interrupt_main](https://github.com/mcepl/cpython/blob/7e414f2f495e73258f55bdef153bc19fb8a9e472/Modules/_threadmodule.c#L1106) and I am really not sure what's the state of GIL there. Of course, the line in the comment should go.

**Comment by mcepl** [comment]
@vstinner What's missing from this PR to be merged?

File: Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst]: typo: Victor

**Comment by mcepl** [comment]
@pitrou Could I ask for a review, here (and/or merge)?

File: Doc/c-api/exceptions.rst
Comment (pitrou) [inline Doc/c-api/exceptions.rst]: "Deprecated version of ...".

File: Doc/c-api/exceptions.rst
Comment (pitrou) [inline Doc/c-api/exceptions.rst]: "... and ``-1`` is returned. ``0`` is returned on success."

File: Doc/c-api/exceptions.rst
Comment (pitrou) [inline Doc/c-api/exceptions.rst]: 3.8, no?

File: Lib/test/test_threading.py
Comment (pitrou) [inline Lib/test/test_threading.py]: Please leave a space after `#` and before the comment.

File: Lib/test/test_threading.py
Comment (pitrou) [inline Lib/test/test_threading.py]: s/threat/thread/

File: Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst
Comment (pitrou) [inline Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst]: Victor, not Viktor

File: Doc/c-api/exceptions.rst
Comment (pitrou) [inline Doc/c-api/exceptions.rst]: That doesn't sound right. `PyErr_SetString` probably requires the GIL to be held.
If not requiring the GIL is desirable, then errors should be ignored rather than reported.

File: Doc/library/_thread.rst
Comment (pitrou) [inline Doc/library/_thread.rst]: 3.8

**Review by pitrou — CHANGES_REQUESTED** [review]
I have a problem with this PR. It doesn't fix `PyErr_SetInterrupt` at all, it requires the caller to switch to another API. I think it would make sense to fix `PyErr_SetInterrupt` so that it does nothing if SIGINT isn't handled by Python.

File: Doc/c-api/exceptions.rst
Comment (mcepl) [reply Doc/c-api/exceptions.rst]: @pitrou I don't understand what exactly you object to here. Could you please elaborate?

File: Lib/test/test_threading.py
Comment (mcepl) [reply Lib/test/test_threading.py]: That sounds like a very much Freudian typo ;).

File: Doc/c-api/exceptions.rst
Comment (pitrou) [reply Doc/c-api/exceptions.rst]: The missing "of" in the sentence above.

File: Modules/signalmodule.c
Comment (mcepl) [reply Modules/signalmodule.c]: Right. The comment is gone.

File: Doc/c-api/exceptions.rst
Comment (mcepl) [reply Doc/c-api/exceptions.rst]: Right. Fixed.

File: Doc/c-api/exceptions.rst
Comment (mcepl) [reply Doc/c-api/exceptions.rst]: Comment removed.

**Comment by mcepl** [comment]
> I have a problem with this PR. It doesn't fix ``PyErr_SetInterrupt`` at all, it requires the caller to switch to another API. I think it would make sense to fix ``PyErr_SetInterrupt`` so that it does nothing if ``SIGINT`` isn't handled by Python.

@pitrou Would you be OK with this pull request, if instead of creating new function, we would just change ``PyErr_SetInterrupt`` to the proposed reading of ``PyErr_SetInterruptWithErr`` as it stands right now and not create new function at all?

**Comment by pitrou** [comment]
> we would just change PyErr_SetInterrupt to the proposed reading of PyErr_SetInterruptWithErr as it stands right now and not create new function at all?

Do you mean raise an error or not? I don't think it's worthwhile to raise an error, TBH. These are low-level API functions that are generally not a good idea anyway. @vstinner may disagree.

**Comment by mcepl** [comment]
> Do you mean raise an error or not? I don't think it's worthwhile to raise an error, TBH. These are low-level API functions that are generally not a good idea anyway. @vstinner may disagree.

I came to this pull request from https://bugzilla.opensuse.org/1094814 , so there is some real reason for it. What do you think should be the ideal solution for that bug? @pitrou @vstinner

**Comment by vstinner** [comment]
This PR was in my TODO list (of PRs that I should review), but sadly I failed to find the bandwidth to review it. Don't hesitate to retry to ping me in two weeks if nobody came up with a review in the meanwhile ;-)

**Comment by plocharz-9livesdata** [comment]
@vstinner ping ;)

**Comment by plocharz-9livesdata** [comment]
We encounter similar problem but in `signal.signal` function instead of `interrupt_main`. This pull request does not seem to solve it, but I was wondering if the source of the problem is the same?

Following code can reproduce the problem:
```python
import signal
import os
import sys
import time

import multiprocessing

def fail():
    def handler(*args):
        pass

    while True:
        signal.signal(signal.SIGUSR1, signal.SIG_IGN)
        signal.signal(signal.SIGUSR1, handler)

proc = multiprocessing.Process(target=fail)
proc.start()
time.sleep(1)
pid = proc.pid

i = 0
try:
    while proc.is_alive():
        os.kill(pid, signal.SIGUSR1)
        time.sleep(0.001)
finally:
    proc.join()
```
Tested on this branch, it fails with:
```
Process Process-1:
Traceback (most recent call last):
  File "/home/plocharz/projects/cpython/Lib/multiprocessing/process.py", line 302, in _bootstrap
    self.run()
  File "/home/plocharz/projects/cpython/Lib/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "keep-killing.py", line 13, in fail
    signal.signal(signal.SIGUSR1, signal.SIG_IGN)
  File "/home/plocharz/projects/cpython/Lib/signal.py", line 47, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
TypeError: 'int' object is not callable
```

**Comment by mcepl** [comment]
@serhiy-storchaka @pablogsal Could I ask for merge here, please? What remains to be fixed?

File: Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst
Comment (mcepl) [reply Misc/NEWS.d/next/Library/2016-07-27-11-06-43.bpo-23395.MuCEX9.rst]: That has been already fixed.

**Comment by mcepl** [comment]
> I have a problem with this PR. It doesn't fix `PyErr_SetInterrupt` at all, it requires the caller to switch to another API. I think it would make sense to fix `PyErr_SetInterrupt` so that it does nothing if SIGINT isn't handled by Python.

@pitrou Is this what you meant?

**Comment by pitrou** [comment]
I don't know if it's ok to change the return type of `PyErr_SetInterrupt`, because of the stable ABI. @vstinner @serhiy-storchaka What do you think?

(otherwise I'm ok with the PR on the principle)

**Comment by mcepl** [comment]
> I don't know if it's ok to change the return type of `PyErr_SetInterrupt`, because of the stable ABI. @vstinner @serhiy-storchaka What do you think?
> 
> (otherwise I'm ok with the PR on the principle)

Right, it is possible not to return anything and just to rely on exceptions to report exceptional state, but I thought that returning errorlevel is more standard C way of doing things. You decide.

**Comment by pitrou** [comment]
My original suggestion was to do nothing if there is no SIGINT signal handler. Don't raise an exception, don't print a warning.

By the way, it may be that some people rely on calling `PyErr_SetInterrupt` without the GIL.

**Comment by mcepl** [comment]
> My original suggestion was to do nothing if there is no SIGINT signal handler. Don't raise an exception, don't print a warning.

Hmm, that would do, but it seems unkind not to tell users something is wrong.

> Edit: `PyErr_SetInterrupt` can be called without the GIL (see docs), so we can't raise an exception or emit a warning.

So, in the end we have to return errorlevel to signal anything?

**Comment by pitrou** [comment]
> So, in the end we have to return errorlevel to signal anything?

If we want to return an error, yes, we need a new API. But the existing API should still be fixed.

**Comment by mcepl** [comment]
> > So, in the end we have to return errorlevel to signal anything?
> 
> If we want to return an error, yes, we need a new API. But the existing API should still be fixed.

OK, so this one? Yes, we may discuss some alternative API later, but I would love to get this into 3.8 so let's keep it as simple as possible.

**Comment by pitrou** [comment]
On the principle, yes. The PR needs polishing, I'll do it.

**Review by pitrou — APPROVED** [review]

**Comment by mcepl** [comment]
> On the principle, yes. The PR needs polishing, I'll do it.

You are right, I forgot ``try ... finally``. That's handy.

**Comment by mcepl** [comment]
> On the principle, yes. The PR needs polishing, I'll do it.

You are right, I forgot ``try ... finally``. That's handy.

**Pull request merged**
